src/oss/langchain/streaming.mdx

```diff
@@ -509,6 +511,386 @@ for await (const [streamMode, chunk] of await agent.stream(
+
+
+def _process_stream(stream_mode: str, data: Any) -> None:
+    if stream_mode == "messages":
+        token, metadata = data
+        if (
+            metadata["langgraph_node"] == "model"
+            and isinstance(token, AIMessage)  # help type-checking
+        ):
+            _process_message_chunk(token)  # [!code highlight]
+    if stream_mode == "updates":
+        for source, update in data.items():
+            if source in ("model", "tools"):
+                _process_completed_message(source, update["messages"][-1])  # [!code highlight]
+
+
+input_message = {"role": "user", "content": "What is the weather in Boston?"}
+for stream_mode, data in agent.stream(
+    {"messages": [input_message]},
+    stream_mode=["messages", "updates"],  # [!code highlight]
+):
+    _process_stream(stream_mode, data)
+```
+```shell title="Output" expandable
+[{'name': 'get_weather', 'args': '', 'id': 'call_D3Orjr89KgsLTZ9hTzYv7Hpf', 'index': 0, 'type': 'tool_call_chunk'}]
+[{'name': None, 'args': '{"', 'id': None, 'index': 0, 'type': 'tool_call_chunk'}]
+[{'name': None, 'args': 'city', 'id': None, 'index': 0, 'type': 'tool_call_chunk'}]
+[{'name': None, 'args': '":"', 'id': None, 'index': 0, 'type': 'tool_call_chunk'}]
+[{'name': None, 'args': 'Boston', 'id': None, 'index': 0, 'type': 'tool_call_chunk'}]
+[{'name': None, 'args': '"}', 'id': None, 'index': 0, 'type': 'tool_call_chunk'}]
+Tool calls: [{'name': 'get_weather', 'args': {'city': 'Boston'}, 'id': 'call_D3Orjr89KgsLTZ9hTzYv7Hpf', 'type': 'tool_call'}]
+Tool response: [{'type': 'text', 'text': "It's always sunny in Boston!"}]
+The| weather| in| Boston| is| **|sun|ny|**|.|
+```
+
+### Streaming with human-in-the-loop
+
+To handle human-in-the-loop [interrupts](/oss/langchain/human-in-the-loop), we build on the [above example](#streaming-tool-calls):
+
+1. We configure the agent with [human-in-the-loop middleware and a checkpointer](/oss/langchain/human-in-the-loop#configuring-interrupts)
+2. We collect interrupts generated during the `"updates"` stream mode
+3. We respond to those interrupts with a [command](/oss/langchain/human-in-the-loop#responding-to-interrupts)
+
+```python
+from typing import Any
+
+from langchain.agents import create_agent
+from langchain.agents.middleware import HumanInTheLoopMiddleware
+from langchain.messages import AIMessage, AnyMessage
+from langgraph.checkpoint.memory import InMemorySaver
+from langgraph.types import Command, Interrupt
+
+
+def get_weather(city: str) -> str:
+    """Get weather for a given city."""
+
+    return f"It's always sunny in {city}!"
+
+
+checkpointer = InMemorySaver()
+
+agent = create_agent(
+    "openai:gpt-5.2",
+    tools=[get_weather],
+    middleware=[  # [!code highlight]
+        HumanInTheLoopMiddleware(interrupt_on={"get_weather": True}),  # [!code highlight]
+    ],  # [!code highlight]
+    checkpointer=checkpointer,  # [!code highlight]
+)
+
+
+def _process_message_chunk(token: AIMessage) -> None:
```

Comment:
Are we able to remove the _process_* helpers as they act like sinks into print but usually users will not want a sync but to rewrite into a different generator